Remove insertions and deletions to deprecated tunnel map

pippolo84 opened the pull request:
Tunnel map has been deprecated in favor of the ipcache map in https://github.com/cilium/cilium/pull/38483. Since we recreate the tunnel map at each agent restart, the insertions and deletions of the node allocation CIDRs into the tunnel map are not needed for backward compatibility.

The PR removes the stale code in the linux node manager, leaving just the bpf tunnel map to ease the agent upgrade and downgrade processes. Moreover, the PR extends the node manager unit tests to have the same coverage we had with the previous tunnel map based implementation.

~Blocked waiting for https://github.com/cilium/cilium/pull/38483 to be merged first.~

pippolo84 committed:
node/manager: Add test infrastructure for ipcache metadata

The current unit tests for the node manager lacks the ability to check
the metadata added or removed through the IPCache API. The commit
extends the mocked ipcache in order to propagate the metadata associated
to each upserted or removed prefixes. Moreover, it extends the
expectIPCacheUpdate to check those metadata, but only when they are
expected, in order to keep the helper backward compatible with its
usages in already existing tests.

Apart from the change in the infrastructure, this commit does not affect
the actual tests logic and thus their coverage.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
node/manager: Extend unit tests for multiple allocation CIDRs

Extend the current unit tests to check the ipcache metadata upserted
and removed in case of node multiple allocation CIDRs. Both the
scenarios with (TestNodeEncryption) and without (TestIpcache) encryption
are verified.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

doniacld commented on pkg/node/manager/manager_test.go:
```diff
@@ -74,23 +75,23 @@ func AddrOrPrefixToIP(ip string) (netip.Prefix, error) {
 }
 
-func (i *ipcacheMock) Upsert(ip string, hostIP net.IP, hostKey uint8, k8sMeta *ipcache.K8sMetadata, newIdentity ipcache.Identity) (bool, error) {
+func (i *ipcacheMock) Upsert(ip string, hostIP net.IP, hostKey uint8, k8sMeta *ipcache.K8sMetadata, newIdentity ipcache.Identity, aux ...ipcache.IPMetadata) (bool, error) {
```
Dumb question but `aux` naming is for auxiliary?

pippolo84 replied:
That's correct, it stands for "auxiliary" metadata :+1:

doniacld reviewed (approved):
Nice work 🧹

julianwiedmann commented on pkg/datapath/linux/node_linux_test.go:
```diff
@@ -422,281 +398,6 @@ func (s *linuxPrivilegedBaseTestSuite) TestAuxiliaryPrefixes(t *testing.T) {
-		require.NoError(t, err)
-		require.True(t, underlayIP.Equal(externalNodeIP2))
-
-		foundRoute, err := linuxNodeHandler.lookupNodeRoute(ip4Alloc1, false)
-		require.NoError(t, err)
-		require.NotNil(t, foundRoute)
-	}
```
I understand that we can drop the testing for the tunnel map. But why not keep the testing for the routes?

pippolo84 replied:
I was under the impression that `TestUpdateNodeRoute` was enough but you're right, `TestNodeUpdateEncapsulation` and `TestNodeUpdateEncapsulationWithOverride` exercise the exported methods `NodeUpdate` and `NodeDelete`, so they guarantee a better coverage.
I removed the steps in the above subtests that were redundant (we don't need to test changes related to tunnel map now) but kept the interesting ones. Finally, I removed `TestAgentRestartOptionChanges` that was related to the tunnel map only. :+1:

julianwiedmann replied:
Could you please split the patch in two - one that removes *only* the tunnel map-related parts, and a second one that subsequently removes all the other redundant testing (with an explanation why they are redundant) ? 

This will make review much more obvious, and allows us to easily roll back if the tests turn out not to be redundant after all ;).

pippolo84 replied:
Splitted the commit in multiple ones, to address each separate case differently. I've also realized there was a mistake in the current test suite that led to the execution of the same subtest twice (while another one was not called at all). Please review commit by commit for all the info :pray:

julianwiedmann replied:
thank you! :heart:

pippolo84 committed:
node: Improve node update direct routing test

Add a step to verify that, in case of a node deletion, all the routes
are removed, even the one related to the secondary CIDRs.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
node: Remove insertions and deletions to/from tunnel map

Given that the datapath now relies on the ipcache map for tunneling, it
is possible to remove the insertions and deletions of allocations CIDRs
into the deprecated tunnel map, alongside all the related unit testing.

Keeping the prefixes insertions into the tunnel map is not useful in
case of an agent downgrade either. That's because the agent re-create
the tunnel map at each restart, thus all the data in the tunnel map
inserted by an upgraded agent for backward compatibility would be erased
after the restart of the downgraded agent.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
node: Remove cidrNodeMappingUpdateRequired test

The function is part of the logic to update the deprecated tunnel map.
Specifically it is responsible for returning a list of CIDRs that need
to be updated in the map after a node object update.  Since the
insertions and deletions into the tunnel map are not useful anymore and
they will be removed in a subsequent commit, remove the test.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
node: Rework tests for node update with encap

Tunnel map has been deprecated and all the related logic will be removed
in a subsequent commit. Therefore, update the tests exercising the node
update and delete code paths dropping all the checks related to the map,
but keeping the assertions that verify the installed routes.

Note that the previous version of the "nodev2" step in the test can be
removed altogether, since it was designed to change solely the remote
node address: that had an effect only on the tunnel map entries, not on
the routes that point to the host device.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
node: Rework the agent restart test

TestAgentRestartOptionChanges is designed to verify the restart of the
linux node manager with a different set of options, specifically
changing the enabling of the v4 and v6 address families.

Since the tunnel map is deprecated, update the test to verify the
correct update of the routes instead.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
node: Fix node validation direct routing test case

The call to the TestNodeValidationDirectRouting case was inadvertently
removed in a previous commit. Specifically, when the new TestAll was
added, the "TestNodeValidationDirectRouting" subtest case was
incorrectly set to call another subtest case
("TestAgentRestartOptionChanges").

Fixes: a7d4e63f37 ("pkg/datapath: Replace gocheck with built-in go test")

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>
